### PR TITLE
Call `optimizer.zero_grad()` before backward inside closure in AutoOpt

### LINKED ISSUE #4083: using LBFGS optimizer in pytorch lightening the model is not converging as compared to native pytoch + LBFGS
### Common bugs:
Comparing the results of LBFGS + Pytorch lightening to native pytorch + LBFGS, Pytorch lightening is not able to update wights and  model is not converging. there are some issues to point out:
1. Adam + Pytorch lightening on MNIST works fine, however LBFGS + Pytorch lightening is not working as expected.
2. LBFGS + Native pytorch works very well, however when we try LBFGS + Pytorch lightening it does not work as expected.

## 🐛 Bug

LBFGS + Pytorch Lightening has problem converging and weights are updating as compared to Adam + Pytorch lightening.

#### Code sample
```
import os
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torchvision.datasets import MNIST
from torchvision import transforms,datasets
from torch.utils.data import DataLoader,random_split
import pytorch_lightning as pl 
from IPython.display import clear_output

class LightningMNISTClassifier(pl.LightningModule):
  def __init__(self):
    super(LightningMNISTClassifier,self).__init__()
    self.layer_1 = nn.Linear(28 * 28, 128)
    self.layer_2 = nn.Linear(128, 256)
    self.layer_3 = nn.Linear(256, 10)
    
  def forward(self, x):
    batch_size, channels, width, height = x.size()
    x=x.view(batch_size,-1)
    # layer 1
    x = self.layer_1(x)
    x = torch.relu(x)
    # layer 2
    x = self.layer_2(x)
    x = torch.relu(x) 
    # layer 3
    x = self.layer_3(x)
    # probability distribution over labels
    x = torch.log_softmax(x, dim=1)  
    return x 
  def prepare_data(self):
    transform=transforms.Compose([transforms.ToTensor(), 
                                  transforms.Normalize((0.1307,), (0.3081,))])
    # prepare transforms standard to MNIST
    mnist_train = MNIST(os.getcwd(), train=True, download=True, transform=transform)
    mnist_test = MNIST(os.getcwd(), train=False, download=True, transform=transform)  
    self.mnist_train, self.mnist_val = random_split(mnist_train, [55000, 5000])

  def train_dataloader(self):
    return DataLoader(self.mnist_train,batch_size=1024)
 
  # def val_dataloader(self):
  #   return DataLoader(self.mnist_val,batch_size=1024)
  # def test_dataloader(self):
  #   return DataLoader(self.mnist_test,batch_size=1024)


  def configure_optimizers(self):
    # optimizer=optim.Adam(self.parameters(),lr=1e-3)
    optimizer = optim.LBFGS(self.parameters(), lr=1e-2)
    return optimizer

  # def backward(self, trainer, loss, optimizer):
  #   loss.backward(retain_graph=True)


  def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx,
                     second_order_closure, on_tpu=False, using_native_amp=False,
                     using_lbfgs=False):
        # update params
      optimizer.step(second_order_closure) 

  def cross_entropy_loss(self,logits,labels):
    return F.nll_loss(logits,labels)

  def training_step(self,train_batch,batch_idx):
    x,y=train_batch
    logits=self.forward(x)
    loss=self.cross_entropy_loss(logits,y)
    return  {'loss':loss}

  def training_epoch_end(self,outputs):
    avg_loss=torch.stack([x['loss'] for x in outputs]).mean()
    print('epoch={}, avg_Train_loss={:.2f}'.format(self.current_epoch,avg_loss.item()))
    # return {'avg_train_loss':avg_loss}

  # def validation_step(self,val_batch,batch_idx):
  #   x,y=val_batch
  #   logits=self.forward(x)
  #   loss=self.cross_entropy_loss(logits,y)
  #   return {'val_loss':loss}
  # def validation_epoch_end(self,outputs):
  #   avg_loss=torch.stack([x['val_loss'] for x in outputs]).mean()
  #   print('epoch={}, avg_Test_loss={:.2f}'.format(self.current_epoch,avg_loss.item()))
  #   return {'avg_val_loss':avg_loss}

model=LightningMNISTClassifier()
#from pytorch_lightning.callbacks import EarlyStopping
trainer=pl.Trainer(max_epochs=400,gpus=1,
                  #  check_val_every_n_epoch=2,
                  #  accumulate_grad_batches=5,
#                   early_stop_callback=early_stop,
                  #  limit_train_batches=50,
#                   val_check_interval=0.25,
                   progress_bar_refresh_rate=0,
#                   num_sanity_val_steps=0,
                   weights_summary=None)
clear_output(wait=True)
trainer.fit(model)Preformatted text.

```
### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

Environment:
-Colab and pycharm 
-PyTorch version: 1.6.0+CPU and GPU
-pytorch-lightning==1.0.0rc3

### LINKED ISSUE #5545: GAN Manual optimization not working after 1.0.7
## 🐛 Bug

I have a GAN model and as of 1.0.7 it no longer trains correctly. I've tried to troubleshoot the issue to the best of my ability but I have no idea what's causing the problem.

## Please reproduce using the BoringModel

https://colab.research.google.com/gist/import-antigravity/0730243bb11b56031110fd6aa7d58971/the-boringmodel.ipynb

### To Reproduce

See boringmodel above^^^

### Expected behavior

Using the colab notebook switch between version 1.0.6 and 1.0.7 to see the bug. For 1.0.6 after training a few epochs it's clear the GAN is beginning to converge, for 1.0.7 it just makes noise.

### Environment

 - PyTorch Version (e.g., 1.0): 1.0.7
 - OS (e.g., Linux): macOS, Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): n/a
 - Python version: 3.8
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Fixes #4083
Fixes #5545
To-check #6134

### Description of the changes
Makes sure to call `zero_grad` inside the closure function (`TrainerLoop.training_step_and_backward()`).

Note that this positions the `zero_grad` call before `backward`, as generally suggested throughout [PyTorch's docs](https://pytorch.org/tutorials/beginner/pytorch_with_examples.html).

Reported that LBFGS doesn't work In #4083, we then found that the number of times `zero_grad` is actually called is different between Lightning and pure PyTorch:
- Lightning calls `closure` 20 times and `zero_grad` only 1 time while
- PyTorch calls `closure` 20 times and `zero_grad` 20 times where 20 is the value of `torch.optim.LBFGS(..., max_iter=20)`.  (because obviously `closure` calls `zero_grad` inside. See the sample scripts below)

As mentioned in [the PyTorch docs](https://pytorch.org/docs/stable/optim.html#optimizer-step-closure), the closure should call `zero_grad`, but the current Lightning calls it outside the closure not inside, and thus it's not working properly when using optimizers which need re-evaluation of the loss in `optimizer.step(closure)`.
> The closure should clear the gradients, compute the loss, and return it.

### TODO
- [x] Call `zero_grad` in closure (and remove `zero_grad` calls outside the closure)
- [x] Update docs to reflect the new `zero_grad` position
- [ ] ~~Update docs to recommend using manual optimization when using a similar optimizer to `torch.optim.LBFGS` which needs reevaluation of the loss via closure.~~
- [ ] ~Ensure that scheduler.step is called the same number of times as optimizer.step in manual optimization~ I'll disable `scheduler.step` in manual optimization in another PR. cc: @carmocca

Here are the minimal code examples using BoringModel.
<details>
<summary>Lightning code</summary>

```python
import pytorch_lightning as pl
import torch
from torch.utils.data import DataLoader, Dataset
pl.seed_everything(42)

class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)
    def __getitem__(self, index):
        return self.data[index]
    def __len__(self):
        return self.len

class BoringModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
    def forward(self, x):
        return self.layer(x)
    def loss(self, batch, prediction):
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))
    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}
    def training_step_end(self, training_step_outputs):
        loss = training_step_outputs["loss"]
        print("loss:", loss.item())
        return training_step_outputs
    def configure_optimizers(self):
        # optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        optimizer = torch.optim.LBFGS(self.parameters(), lr=0.01, max_iter=20)
        return optimizer

def main():
    ds = RandomDataset(32, 100000)
    dl = DataLoader(ds, batch_size=1024)
    model = BoringModel()
    trainer = pl.Trainer(
        progress_bar_refresh_rate=0,
        fast_dev_run=1,
    )
    trainer.fit(model, dl)

if __name__ == "__main__":
    main()
```

</details>

<details>
<summary>Pure PyTorch code</summary>

```python
import torch
import torch.nn as nn
from pytorch_lightning import seed_everything
from torch.utils.data import DataLoader, Dataset
seed_everything(42)

class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)
    def __getitem__(self, index):
        return self.data[index]
    def __len__(self):
        return self.len

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.layer = torch.nn.Linear(32, 2)
    def forward(self, x):
        return self.layer(x)

def main():
    ds = RandomDataset(32, 100000)
    dl = DataLoader(ds, batch_size=1024)
    model = Model()
    # optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    optimizer = torch.optim.LBFGS(model.parameters(), lr=0.01, max_iter=20)
    for epoch in range(3):
        for i, x in enumerate(dl):
            def closure():
                prediction = model(x)
                loss = torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))
                optimizer.zero_grad()  # removing this line causes the same bug as in Lightning script
                loss.backward()
                print("loss:", loss.item())
                return loss
            loss_out = optimizer.step(closure=closure)

if __name__ == '__main__':
    main()
```

</details>

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

cc: @carmocca @tchaton

### CODE REVIEW

File: docs/source/common/optimizers.rst [hunk 1]
```diff
@@ -23,32 +23,31 @@ to manually manage the optimization process. To do so, do the following:
...
 
-
 .. code-block:: python
 
     def training_step(batch, batch_idx, optimizer_idx):
         opt = self.optimizers()
 
         loss = self.compute_loss(batch)
         self.manual_backward(loss)
-        opt.step()
```
Comment (akihironitta) [inline docs/source/common/optimizers.rst:44]: This `opt.step()` is placed at the wrong place, isn't it?

Comment (akihironitta) [inline docs/source/common/optimizers.rst:44]: This `opt.step()` is placed at the wrong place, isn't it?

File: docs/source/common/optimizers.rst [hunk 2]
```diff
@@ -300,7 +304,7 @@ override the :meth:`optimizer_step` function.
 
 For example, here step optimizer A every 2 batches and optimizer B every 4 batches
 
-.. note:: When using Trainer(enable_pl_optimizer=True), there is no need to call `.zero_grad()`.
```
Comment (akihironitta) [inline docs/source/common/optimizers.rst:303]: `enable_pl_optimizer` was removed in #6163.

Comment (carmocca) [reply docs/source/common/optimizers.rst:303]: You can just merge master

File: docs/source/common/optimizers.rst (hunk 1 above)
Comment (carmocca) [reply docs/source/common/optimizers.rst:44]: yes

File: docs/source/common/optimizers.rst (hunk 2 above)
Comment (akihironitta) [reply docs/source/common/optimizers.rst:303]: Ah, right, I thought it was missed. Will merge master.

**Review by justusschock — APPROVED** [review]

**Comment by carmocca** [comment]
Great work @akihironitta! I pushed a commit refactoring the tests since there was lot of duplication.

Do you think we can add/modify one test to assert the call order for `step` and `zero_grad` in automatic optimization?

**Comment by akihironitta** [comment]
@carmocca Yes, I'll make sure to add it within a few hours.

**Review by tchaton — APPROVED** [review]
Great work there !

**Comment by akihironitta** [comment]
@carmocca I added the test similarly to [the existing test](https://github.com/PyTorchLightning/pytorch-lightning/blob/85431ffaa4618dbeeb0fe8fd0216948f19c8f118/tests/models/test_hooks.py#L265), so mind having a look again?

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -1789,3 +1693,76 @@ def training_step(self, batch, batch_idx):
...
     trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, gpus=1)
     trainer.fit(model, train_data)
+
+
+def test_train_loop_system(tmpdir):
+    """
+    Test the following methods are called in the order in automatic optimization.
+    1. training_step
+    2. zero_grad (skip when accumulate gradients)
+    3. backward
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1703]: Great, can you add optimizer.step too?

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -1789,3 +1693,76 @@ def training_step(self, batch, batch_idx):
...
+        max_epochs=1,
+        limit_train_batches=3,
+        limit_val_batches=1,
+        limit_test_batches=1,
+        progress_bar_refresh_rate=0,
+    )
+
+    class TestOptimizer(SGD):
+        def zero_grad(self, *args, **kwargs):
+            called_methods.append(inspect.currentframe().f_code.co_name)
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1718]: You can set the name yourself, no need to use the current frame

```suggestion
            called_methods.append("zero_grad")
```

same for the other occurrences

File: tests/trainer/test_trainer.py (hunk 3 above)
Comment (akihironitta) [reply tests/trainer/test_trainer.py]: @carmocca The test has been updated.

**Review by carmocca — APPROVED** [review]

File: docs/source/common/optimizers.rst [hunk 5]
```diff
@@ -137,8 +135,12 @@ Here is an example on how to use it:
...
 Automatic optimization
 ======================
-With Lightning most users don't have to think about when to call .backward(), .step(), .zero_grad(), since
-Lightning automates that for you.
+With Lightning most users don't have to think about when to call ``.zero_grad()``, ``.backward()`` and ``.step()``
+since Lightning automates that for you.
+
+.. warning::
+   Before 1.2.1, ``.zero_grad()`` was called after ``.backward()`` and ``.step()`` internally.
+   From 1.2.1, Lightning calls ``.zero_grad()`` before ``.backward()``.
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:143]: isn't it better to move this into 1.3? @carmocca 
also 1.2.1 is already out, so this message must be updated anyway

Comment (justusschock) [reply docs/source/common/optimizers.rst:143]: @awaelchli I don't think so, since this is actually a bugfix to finally support lbfgs

File: CHANGELOG.md [hunk 6]
```diff
@@ -17,6 +17,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
+- Changed the order of `zero_grad` and `backward` calls in automatic optimization. Lightning now calls `zero_grad` before `backward` ([#6147](https://github.com/PyTorchLightning/pytorch-lightning/pull/6147))
+
```
[the remark this answers is not in the document]
Comment (akihironitta) [reply CHANGELOG.md:21]:
```suggestion
- Changed the order of `backward`, `step`, `zero_grad` to `zero_grad`, `backward`, `step` ([#6147](https://github.com/PyTorchLightning/pytorch-lightning/pull/6147))

```

File: docs/source/common/optimizers.rst (hunk 5 above)
Comment (akihironitta) [reply docs/source/common/optimizers.rst]: @awaelchli Updated to 1.2.2, but I'll update it to 1.3 if needed.

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (akihironitta) [reply CHANGELOG.md]: @awaelchli I applied the my above suggestion, how does it look now?

Let me note that the statement is fine for general users, but for some users who care what's really going on inside, the statement is actually false because `step` calls `zero_grad` and `backward` and updates params inside.

[the remark this answers is not in the document]
Comment (awaelchli) [reply CHANGELOG.md]: Great. 

It just looks like that because of the closure, but the actual parameter update (step) comes at the end, so I don't see a problem.

File: docs/source/common/optimizers.rst (hunk 5 above)
Comment (carmocca) [reply docs/source/common/optimizers.rst]: > this is actually a bugfix to finally support lbfgs

And (some?) GANs
